capsules/src/usb/descriptors.rs

```diff
@@ -370,6 +375,183 @@ impl Descriptor for DeviceDescriptor {
     }
 }
 
+/// Buffer for holding the device descriptor.
+// TODO it's dumb that these are Cells, but doing otherwise would require
+// rewriting the `write_to` functions
+pub struct DeviceBuffer {
+    pub buf: [Cell<u8>; 19],
+    pub len: usize,
+}
+
+impl DeviceBuffer {
+    pub fn write_to(&self, buf: &[Cell<u8>]) -> usize {
+        for i in 0..self.len {
+            buf[i].set(self.buf[i].get());
+        }
+        self.len
+    }
+}
+
+/// Buffer for holding the configuration, interface(s), and endpoint(s)
+/// descriptors. Also includes class-specific functional descriptors.
+pub struct DescriptorBuffer {
+    pub buf: [Cell<u8>; 128],
+    pub len: usize,
+}
```

Comment:
From a memory usage point of view, I find it unfortunate to store some descriptors (e.g. HID) twice: one time already serialized here, and another time in the `ClientCtrl` object. Can't we serialize everything on-demand in the `handle_standard_device_request` function, as before?

Reply:
Would you be willing to add a commit? This code works, and gets us closer to a long standing item on our wishlist. The USB stack is by no means final, and I'm sure there are optimizations possible.

Reply:
> Would you be willing to add a commit?

As far as I'm aware, I'm not able to push commits in this branch, as I'm not a "reviewer with write access".

Reply:
I think serializing everything immediately is actually the way to go. The ClientCtrl doesn't actually need to know much about the descriptors other than just hang on to several buffers of memory and provide them when desired. So it doesn't need a full struct sitting around. A way to stop the duplication would be to have separate buffers for endpoints, interfaces, and other stuff (like the HID) and then combine buffers when things are requested by the host. That would allow you to have only one copy of the HID. We didn't go that full route because the HID descriptor isn't very large so just having two copies lying around doesn't seem to matter much.

The other reason to go with a serialized approach is that it was getting difficult to make lifetimes work for the descriptors. The ClientCtrl can't have structs that it copies fields into because it doesn't know which descriptors are going to exist or how many. It's way easier for the ClientCtrl to have an array of bytes with a fixed size that it can just copy descriptors into than having pointers to arrays of structs that you have to reason about the lifetimes of. I'm sure there's some way to get things working that way, but I couldn't figure it out.

Reply:
If you think this is really important to change before this PR goes in and you have some ideas about how to do so, I think you could pull this branch into your own Tock repo, make changes, and then make a PR against this branch rather than against master.

Reply:
> If you think this is really important to change before this PR goes in and you have some ideas about how to do so, I think you could pull this branch into your own Tock repo, make changes, and then make a PR against this branch rather than against master.

It's not really practical to make a PR against another PR to be honest...